bluetooth: host: att: Disconnect on ATT timeout

PavelVPV opened the pull request:
The timeout state is local and can block new ATT operations, but does
not affect the remote side. Disconnecting the GATT connection upon ATT
timeout simplifies error handling for developers. This reduces rare
failure conditions to a common one, without needing special cases for
ATT timeouts.

This PR adds a bsim test to cover new behavior.

Also, added error check when calling `bt_l2cap_send_pdu` (see commit message) which was noticed from the the test logs.

PavelVPV committed:
bluetooth: host: att: Implement disconnect on ATT timeout

The timeout state is local and can block new ATT operations, but does
not affect the remote side. Disconnecting the GATT connection upon ATT
timeout simplifies error handling for developers. This reduces rare
failure conditions to a common one, without needing special cases for
ATT timeouts.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
bluetooth: host: l2cap: Check conn state before queueing PDU

In case of UATT, if a connection was lost while user was holding a read
or write attribute callback, `bt_l2cap_send_pdu` (called from
`att.c::chan_send`) will anyway queue a PDU and trigger tx work. The PDU
won't be sent eventually, but neither will hold an error code, which
will allow it to bypass the error check in `att_on_sent_cb` and call
`att_sent` function.

For EATT `bt_l2cap_chan_send` is used which already handles this case
and the error code is passed to `att_on_sent_cb`.

This change adds connection state check to `bt_l2cap_send_pdu`
preventing from unnecessary code execution when connection does not
exist anymore.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

aescolar commented on tests/bsim/bluetooth/host/att/timeout/prj.conf:
```diff
@@ -0,0 +1,25 @@
+CONFIG_ASSERT=y
```
This seems to be the exact same configuration as tests/bsim/bluetooth/host/att/long_read/prj.conf
Note that for most tests we spend more time building them than actually running them.
If the .conf is the same, is there any reason to not build these as two tests that are part of the same image?

PavelVPV replied:
I am fully agree with you. But the pattern has already been defined. I also see the duplicate code has been used in other tests. But I think it is a separate task to get rid of duplicate code and merge tests into a single image. I saw there was some work done towards this, but seems was not finished.

@alwa-nordic, what do you think? Do you have a task in your backlog for this?

alwa-nordic replied:
Let's leave it out of scope of this PR and create an issue for it. I started some work on a combined build a while ago: https://github.com/zephyrproject-rtos/zephyr/pull/77122, and I would like it done.

alwa-nordic replied:
https://github.com/zephyrproject-rtos/zephyr/issues/79512

aescolar commented on tests/bsim/bluetooth/host/att/timeout/test_scripts/run.sh:
```diff
@@ -0,0 +1,22 @@
+
+Execute ./${dev_exe} "${args_all[@]}" "${args_dev[@]}" -d=1
+
+Execute ./bs_2G4_phy_v1 "${args_all[@]}" -v=6 -sim_length=200e6
```
Please don't submit tests with increased verbosity (what runs in CI should be at -v=2)

alwa-nordic commented on doc/connectivity/bluetooth/bluetooth-le-host.rst:
```diff
@@ -285,7 +285,7 @@ special cases for ATT timeouts.
 .. image:: img/att_timeout.svg
   :align: center
-   :alt: ATT timeout
+  :alt: ATT timeout
```
Nitpick: This change should be moved to the previous commit.

Thalley commented on doc/connectivity/bluetooth/img/att_timeout.svg:
Great that you are adding a figure :) 

Could I suggest that `ATT_READ` is replaced with `ATT_READ_REQ` and `ATT_RESPONSE` is replaced by `ATT_READ_RSP`, so that they match the actual names of the PDUs?

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3148,6 +3148,13 @@ static void att_timeout(struct k_work *work)
 	 * target device on this ATT Bearer.
 	 */
 	bt_att_disconnected(&chan->chan.chan);
+
+	/* The timeout state is local and can block new ATT operations, but does not affect the
+	 * remote side. Disconnecting the GATT connection upon ATT timeout simplifies error handling
+	 * for developers. This reduces rare failure conditions to a common one, allowing developers
+	 * to handle unexpected disconnections without needing special cases for ATT timeouts.
+	 */
+	bt_conn_disconnect(chan->chan.chan.conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
```
If we are doing the `bt_conn_disconnect` here, do we still need to call to `bt_att_disconnected`?

Thalley replied:
> Disconnect command may fail leaving ATT channel active eventually allowing an application to violate the spec

Should we actually check and log that or something?

Thalley commented on tests/bsim/bluetooth/host/att/timeout/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+zephyr_include_directories(
+  ${BSIM_COMPONENTS_PATH}/libPhyComv1/src/
+  ${BSIM_COMPONENTS_PATH}/libUtilv1/src/
+  ../long_read/
```
Somewhat related to https://github.com/zephyrproject-rtos/zephyr/pull/79501/files#r1790165096: If we are using things from `long_read`, do we need a separate build for this?

Would it make sense to move whatever we need from `long_read` to the `testlib`?

Thalley replied:
Looking further, it appears some of it is related to the backchannel syncs? Have you looked at the babblekit to see if we can just use that for this? 

If you need something more than what it supports, maybe it can be added to the babblekit?

Thalley commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,275 @@
+#define CENTRAL_DEVICE_NBR    0
+#define PERIPHERAL_DEVICE_NBR 1
+
+#define ATT_TIMEOUT_SEC 30
```
Arguably we could move `BT_ATT_TIMEOUT` to `att.h` from `att_internal.h` and use that instead of redefining it here

Thalley commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,275 @@
+	 * 1. ATT timeout.
+	 * 2. Successful read.
+	 */
+	for (int i = 0; i < 2; i++) {
```
Hmm, this loop looks a bit weird to me. 

It contains both `if(peripheral)` and `if(central)` stuff, as well as `if i == 0` and `if i == 1`.

Have you considered defining the behavior as functions? e.g. something along the lines of

```c
test_timeout_peripheral(...)
{
    test_peripheral_timeout();
    test_peripheral_read();
}

test_timeout_central
{
    test_central_timeout();
    test_central_read();
}

static void test_timeout(enum bt_att_chan_opt bearer)
{
	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);

	if (central) {
		test_timeout_central();
	} else if (peripheral) {
		test_timeout_peripheral();
	}

}
```

Which gives a better overview than a long loop with several switch cases and very little common code.

PavelVPV replied:
Tried to simplify a bit. But kept the loop. This way it is easier to see what happens on each side at each test step as well as when sync point is called. Ultimately, this is a test, not the final application.

alwa-nordic replied:
I agree with Emil. This loop is a bad abstraction. The `if (timeout) { bt_testlib_conn_unref(&conn); }` is really spaghetti. It's relying on `timeout` being true only in the last iteration of the loop. With this kind of logic, you might as well put the `if (central) {bt_testlib_disconnect}` in the else clause. Why not unroll the loop?

alwa-nordic replied:
Tests are supposed to be simple. Otherwise we will need tests for the tests. :P

PavelVPV replied:
Done

Thalley commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,275 @@
+
+			err = bt_testlib_att_read_by_type_sync(&attr_value, &actual_read_len,
+							       &handle, NULL, conn, bearer,
+							       UUID_2, 1, 0xffff);
```
```suggestion
							       UUID_2, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE);
```

Thalley commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,275 @@
+	}
+
+	__ASSERT(expected_reason == reason, "Unexpected reason %u", reason);
+	bt_conn_unref(conn);
```
Where are the ref's being taken? Do you need to unref here?

PavelVPV replied:
Reference is taken by testlib.

Thalley replied:
So it is. I'm surprised that the testlib doesn't return the reference it took

alwa-nordic replied:
```suggestion
```
This reference is borrowed. You may not move it.

Testlib is indeed not supposed to leak references. Where does this happen?

Thalley replied:
> Testlib is indeed not supposed to leak references. Where does this happen?

The testlib takes a reference in https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/adv.c#L31 for the peripheral and https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/connect.c#L65 for the central

It only calls `bt_conn_unref` if `bt_testlib_conn_wait_free` or `bt_testlib_disconnect` is called, but if neither of those are called, then it leaks the reference. 

Perhaps the `bt_conn_unref` here should be replaced with a call to `bt_testlib_conn_wait_free`?

alwa-nordic replied:
`bt_testlib_adv_conn` moves a to argument `*conn`. `bt_testlib_connect` does the same into `*connp`. Excuse the inconsistent naming of the parameters.

alwa-nordic replied:
The rule to follow is: `bt_conn *`-variables must be initialized to `NULL`. Updating them to `NULL` is only allowed with an accompanying move, e.g. into `bt_conn_unref`. A non-NULL `bt_conn *`-variable must be `NULL` at the end of its lifetime.

PavelVPV replied:
`bt_testlib_connect` calls `bt_conn_le_create` which allocates reference to conn. This reference is dereferences in the disconnect callback.

Thalley replied:
I was also incorrect in saying that `bt_testlib_conn_wait_free` would unreference it. 

But @PavelVPV is correct that if `bt_testlib_disconnect` isn't called, the reference taken by https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/adv.c#L31 or https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/connect.c#L65 is not returned by the testlib in case of uninitiated disconnects (which is the case here where the stack disconnects).

sjanc commented:
#AutoPTS run zephyr GATT/CL/GAT/BV-01-C GATT/CL/GAT/BV-02-C GATT/SR/GAT/BV-01-C GATT/CL/GAT/BV-03-C

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/79501#issuecomment-2397861205, board: nrf52, estimated start time: 22:47:55, test case count: 4, estimated duration: 0:03:32<details><summary>Test cases to be run</summary>GATT/CL/GAT/BV-01-C<br>GATT/CL/GAT/BV-02-C<br>GATT/CL/GAT/BV-03-C<br>GATT/SR/GAT/BV-01-C</details>

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.
<details><summary>Successful tests (4)</summary>GATT    GATT/CL/GAT/BV-01-C             PASS<br>GATT    GATT/CL/GAT/BV-02-C             PASS<br>GATT    GATT/CL/GAT/BV-03-C             PASS<br>GATT    GATT/SR/GAT/BV-01-C             PASS</details>

PavelVPV committed:
tests: bsim: bluetooth: host: att: Add ATT timeout test

This commit tests that the host correctly disconnects from peer when ATT
timeout happens.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

alwa-nordic commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,288 @@
+		}
+
+		if (central) {
+			conn = central_setup(bearer, timeout);
```
This is overwriting a reference. Please initialize `conn` to NULL add `bt_testlib_conn_unref(&conn);` somewhere. Or perform a reference move to some other location.

PavelVPV replied:
No point in setting `conn` to NULL in this function as it will be reassigned here. Removed `bt_conn_unref` from `disconnected` callback.

alwa-nordic commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,288 @@
+	}
+}
+
+static void wait_for_disconnect(bool timeout)
```
I suggest we inline this function for clarity. It's very core and specific to the test.

Thalley commented on tests/bsim/bluetooth/host/att/timeout/main.c:
```diff
@@ -0,0 +1,292 @@
+ * 6. Expect the connection stay alive after a delay equal to ATT timeout.
+ * 7. Expect the central to disconnect.
+ */
+static void test_timeout(enum bt_att_chan_opt bearer)
+{
+	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
+	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);
+	struct bt_conn *conn;
+	int err;
```
Function is much better now, thanks :)

My personal preference would still be to have 2 separate functions (one for central, one for peripheral), since they share so little code, but this is OK

Thalley reviewed (commented):
Do we need an entry in the release notes with this change?

PavelVPV commented:
> Do we need an entry in the release notes with this change?

Added

Thalley commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -107,6 +107,7 @@ Bluetooth
 
   * Added API :c:func:`bt_gatt_get_uatt_mtu` to get current Unenhanced ATT MTU of a given
     connection (experimental).
+  * The host now disconnects from the peer upon ATT timeout.
```
```suggestion
  
  * The host now disconnects from the peer upon ATT timeout.
```

I think you need an empty line to create a new bullet
